ql/src/java/org/apache/hadoop/hive/ql/scheduled/ScheduledQueryExecutionService.java

```diff
@@ -0,0 +1,179 @@
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+public class ScheduledQueryExecutionService implements Closeable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ScheduledQueryExecutionService.class);
+
+  private ScheduledQueryExecutionContext context;
+  private ScheduledQueryExecutor worker;
+
+  public static ScheduledQueryExecutionService startScheduledQueryExecutorService(HiveConf conf) {
+    MetastoreBasedScheduledQueryService qService = new MetastoreBasedScheduledQueryService(conf);
+    ExecutorService executor =
+        Executors.newCachedThreadPool(
+            new ThreadFactoryBuilder().setDaemon(true).setNameFormat("Scheduled Query Thread %d").build());
+    ScheduledQueryExecutionContext ctx = new ScheduledQueryExecutionContext(executor, conf, qService);
+    return new ScheduledQueryExecutionService(ctx);
+  }
+
+  public ScheduledQueryExecutionService(ScheduledQueryExecutionContext ctx) {
+    context = ctx;
+    ctx.executor.submit(worker = new ScheduledQueryExecutor());
+    ctx.executor.submit(new ProgressReporter());
+  }
+
+  static boolean isTerminalState(QueryState state) {
+    return state == QueryState.FINISHED || state == QueryState.ERRORED;
+  }
+
+  class ScheduledQueryExecutor implements Runnable {
+
+    private ScheduledQueryProgressInfo info;
+
+    @Override
+    public void run() {
+      while (true) {
+        ScheduledQueryPollResponse q = context.schedulerService.scheduledQueryPoll();
+        if (q.isSetExecutionId()) {
+          try{
+            processQuery(q);
+          } catch (Throwable t) {
+            LOG.error("Unexpected exception during scheduled query processing", t);
+          }
+        } else {
+          try {
+            Thread.sleep(context.getIdleSleepTime());
+          } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            LOG.info("interrupted");
+            break;
+          }
+        }
+      }
+    }
+
+    public synchronized void reportQueryProgress() {
+      if (info != null) {
+        LOG.info("Reporting query progress of {} as {} err:{}", info.getScheduledExecutionId(), info.getState(),
+            info.getErrorMessage());
+        context.schedulerService.scheduledQueryProgress(info);
+        if (isTerminalState(info.getState())) {
+          info = null;
+        }
+      }
+    }
+
+    private void processQuery(ScheduledQueryPollResponse q) {
+      SessionState state = null;
+      try {
+        HiveConf conf = new HiveConf(context.conf);
+        conf.setVar(ConfVars.HIVE_QUERY_EXCLUSIVE_LOCK, lockNameFor(q.getScheduleKey()));
+        state = SessionState.start(conf);
+        info = new ScheduledQueryProgressInfo();
+        info.setScheduledExecutionId(q.getExecutionId());
+        info.setState(QueryState.EXECUTING);
+        reportQueryProgress();
+        try (
+          IDriver driver = DriverFactory.newDriver(DriverFactory.getNewQueryState(conf), q.getUser(), null)) {
```

Comment:
If HIVE_QUERY_EXCLUSIVE_LOCK is moved out of HiveConf file, you could use conf overlay here by creating a new `getNewQueryState(conf, mapWithOverlay)` method.